### ISSUE TITLE
[Bug]: VllmHiddenStatesGenerator returns silently wrong hidden states with batch_size > 1 or repeated calls

### ISSUE THREAD

**Issue by keenanpepper** [vllm-project/speculators#279]
# [Bug]: VllmHiddenStatesGenerator returns silently wrong hidden states with batch_size > 1 or repeated calls

[repro_speculators_issue.py](https://github.com/user-attachments/files/25118019/repro_speculators_issue.py)

[log.txt](https://github.com/user-attachments/files/25118036/log.txt)

## Summary

`VllmHiddenStatesGenerator.generate()` has two independent bugs that break its documented return contract. The docstring states that `result[i]["hidden_states"]` corresponds to `result[i]["input_ids"]`, but this is violated due to (1) KV cache state leaking between calls and (2) offset-based slicing that doesn't track request identity across scheduler iterations. Both bugs cause **silent data corruption** — no errors are raised, but the hidden states returned are wrong.

This affects anyone using `batch_size > 1` or making multiple calls to the same generator instance, including the intended use case of generating EAGLE training data.

## Environment

- **speculators**: 0.4.0.dev32
- **vLLM**: V1 engine, multiprocess executor, chunked prefill enabled (`max_num_batched_tokens=8192`)
- **Model**: meta-llama/Llama-3.3-70B-Instruct (but issue is model-agnostic)
- **Hardware**: 4× H100 PCIe, `tensor_parallel_size=4`
- **Relevant files**: `speculators/data_generation/vllm_hidden_states_generator.py`, `speculators/data_generation/custom_worker.py`

### Description

After each scheduler iteration, every scheduled request is immediately aborted via `finish_requests(..., FINISHED_ABORTED)`. However, the KV cache blocks allocated for those aborted requests are not fully cleared before the next `generate()` call. This causes the scheduler to believe some prefix tokens are already cached, so it schedules fewer tokens for subsequent calls.

### Observed behavior

Processing 8 sequences individually (one per `generate()` call) on the same generator instance:

```
Sequence lengths: [42, 46, 43, 45, 37, 48, 42, 45]

INDIVIDUAL PROCESSING (ground truth)
  Seq 0: input_len= 42, hs_shape=[42, 8192], match=YES
  Seq 1: input_len= 46, hs_shape=[30, 8192], match=NO (state leak)
  Seq 2: input_len= 43, hs_shape=[27, 8192], match=NO (state leak)
  Seq 3: input_len= 45, hs_shape=[29, 8192], match=NO (state leak)
  Seq 4: input_len= 37, hs_shape=[21, 8192], match=NO (state leak)
  Seq 5: input_len= 48, hs_shape=[32, 8192], match=NO (state leak)
  Seq 6: input_len= 42, hs_shape=[10, 8192], match=NO (state leak)
  Seq 7: input_len= 45, hs_shape=[29, 8192], match=NO (state leak)
```

Only the first call returns the correct number of hidden states. Every subsequent call is short by a multiple of `VLLM_BLOCK_SIZE` (16). The deficit accumulates across calls — each call leaks one block (16 tokens), and by Seq 6 the deficit has grown to 32 tokens (two blocks). The scheduler thinks prefix tokens are already cached from prior aborted requests.

### Description

The `_patched_forward` in `custom_worker.py` captures hidden states as a flat concatenation across all requests in the batch, with no record of which tokens belong to which request. The `generate()` method then slices this concatenated tensor by naive cumulative offset, assuming the tokens appear in input-list order:

```python
offset = 0
for i, seq_len in enumerate(seq_lens):
    layer_states = [h[offset : offset + seq_len] for h in aux_hidden_states]
    offset += seq_len
```

This assumption is violated because:

1. vLLM's chunked prefill scheduler may process sequences across **multiple iterations**, and the order/grouping across iterations doesn't match input order.
2. The scheduler can **reorder sequences** within a batch.
3. The `generate()` method aborts ALL requests after each scheduler iteration, so sequences that were only partially prefilled under chunked prefill **never get fully processed**.

### Observed behavior

Processing the same 8 sequences as a single batch:

```
BATCH PROCESSING (all 8 at once)
  Seq 0: WRONG VALUES
  Seq 1: WRONG SHAPE (got [46, 8192], expected [30, 8192])
  Seq 2: WRONG SHAPE (got [4, 8192], expected [27, 8192])
  Seq 3: EMPTY
  Seq 4: EMPTY
  Seq 5: EMPTY
  Seq 6: EMPTY
  Seq 7: EMPTY

SUMMARY: 5 empty, 3 misaligned, 0 correct out of 8
```

Only 92 of 348 total tokens ended up in the captured hidden states tensor. The scheduler processed a subset of sequences in the first iteration (within the `max_num_batched_tokens=8192` budget), then all requests were aborted. The remaining sequences were never processed at all.

## Impact

These bugs silently corrupt training data for anyone using `VllmHiddenStatesGenerator` with:
- **Multiple calls** to the same generator instance (Bug 1 — cumulative state leak)
- **`batch_size > 1`** (Bug 2 — misalignment and missing sequences)
- **Both** (Bugs 1 + 2 compound)

The offline data generation script (`data_generation_offline.py`) may partially mask these issues depending on batch size and generator reuse patterns, but the API contract is broken regardless.

## Reproduction

See attached `repro_speculators_issue.py` and full output log. The script:
1. Tokenizes 8 prompts of varying length
2. Processes each individually via `generate()` (demonstrates Bug 1 — cumulative state leak)
3. Processes all 8 as a batch (demonstrates Bug 2 — misalignment)
4. Compares shapes and values

Run with: `VLLM_WORKER_MULTIPROC_METHOD=spawn python repro_speculators_issue.py`

Requires access to Llama-3.3-70B-Instruct and 4 GPUs, but the bugs are model-agnostic and would reproduce with any model and `tensor_parallel_size >= 1`.

## Suggested fixes

**Bug 1**: Explicitly free/reset KV cache blocks after aborting requests, or disable prefix caching entirely for the data generation engine since it serves no purpose in this prefill-only workflow.

**Bug 2**: Two sub-issues need fixing:
- **Request tracking**: Tag captured hidden states with request IDs in `_patched_forward` (e.g., capture a list of `(request_id, tensor)` tuples instead of a flat `torch.cat`), then reassemble by request ID in `generate()` rather than by positional offset.
- **Multi-iteration handling**: Don't abort all requests after the first scheduler iteration. Instead, continue running iterations until all requests have been fully prefilled, accumulating hidden states across iterations per request.

**Comment by dsikka** [comment]
cc @shanjiaz
